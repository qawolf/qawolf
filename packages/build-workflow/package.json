{
  "name": "@qawolf/build-workflow",
  "description": "build qawolf workflow from events",
  "version": "0.9.0",
  "license": "BSD-3.0",
  "main": "./lib/index.js",
  "types": "./lib/index.d.ts",
  "files": [
    "lib",
    "src"
  ],
  "scripts": {
    "test": "jest"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/qawolf/qawolf.git"
  },
  "bugs": {
    "url": "https://github.com/qawolf/qawolf/issues"
  },
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@qawolf/browser": "file:../browser",
    "@qawolf/logger": "file:../logger",
    "@qawolf/web": "file:../web"
  },
  "devDependencies": {
    "@qawolf/test": "file:../test",
    "@qawolf/types": "file:../types"
  },
  "peerDependencies": {
    "lodash": "^4.x"
  }
}
