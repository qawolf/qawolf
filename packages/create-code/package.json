{
  "name": "@qawolf/create-code",
  "description": "qawolf: create code",
  "version": "0.9.0",
  "license": "BSD-3.0",
  "main": "./lib/index.js",
  "types": "./lib/index.d.ts",
  "files": [
    "lib",
    "src"
  ],
  "scripts": {
    "test": "jest"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/qawolf/qawolf.git"
  },
  "bugs": {
    "url": "https://github.com/qawolf/qawolf/issues"
  },
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@qawolf/build-code": "file:../build-code",
    "@qawolf/build-workflow": "file:../build-workflow",
    "@qawolf/logger": "file:../logger",
    "@qawolf/repl": "file:../repl"
  },
  "devDependencies": {
    "@qawolf/test": "file:../test",
    "@qawolf/types": "file:../types"
  },
  "peerDependencies": {
    "fs-extra": "^8.x",
    "kleur": "^3.x",
    "lodash": "^4.x"
  }
}
