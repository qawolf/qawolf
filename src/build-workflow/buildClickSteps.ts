import Debug from 'debug';
import { ElementEvent, Step } from '../types';

const debug = Debug('qawolf:buildClickSteps');

const filterClickEvents = (events: ElementEvent[]): ElementEvent[] => {
  return events.filter((event, i) => {
    // track original event index
    (event as any).index = i;

    // ignore system initiated clicks
    if (!event.isTrusted) return false;

    // ignore other actions
    if (!['click', 'mousedown'].includes(event.name)) return false;

    const previousEvent = events[i - 1];
    if (
      previousEvent &&
      ['keydown', 'keyup'].includes(previousEvent.name) &&
      event.time - previousEvent.time < 50
    ) {
      // skip system-initiated clicks triggered by a key press
      // ex. "Enter" triggers a click on a submit input
      debug(`skip click shortly after previous event ${event.time}`);
      return false;
    }

    // ignore clicks on selects
    let name = event.target.name;
    if (name) name = name.toLowerCase();
    return name !== 'select';
  });
};

const groupClickEvents = (
  events: ElementEvent[],
  timeWindow = 200,
): ElementEvent[][] => {
  const groupedEvents = [];
  let group: ElementEvent[] = [];

  events.forEach(event => {
    if (!group.length) {
      // first group
      group.push(event);
      return;
    }

    const lastEvent = group[group.length - 1] as ElementEvent;
    if (event.time - lastEvent.time < timeWindow) {
      // event is within time window
      group.push(event);
      return;
    }

    // start a new group
    groupedEvents.push(group);
    group = [event];
  });

  if (group.length) {
    // append last group
    groupedEvents.push(group);
  }

  return groupedEvents;
};

export const buildClickSteps = (events: ElementEvent[]): Step[] => {
  const clickEvents = filterClickEvents(events);
  const groupedClickEvents = groupClickEvents(clickEvents);
  const steps: Step[] = [];

  groupedClickEvents.forEach(events => {
    let event = events[0] as ElementEvent;

    const inputEvent = events.find(event => {
      const name = event.target.name || '';
      return name.toLowerCase() === 'input';
    });
    if (inputEvent) {
      // if an event in the group is on an  input, assume the click propagated
      // to an element like a checkbox or radio, which is most accurate target
      event = inputEvent;
    }

    steps.push({
      action: 'click',
      index: steps.length,
      event: event,
    });
  });

  return steps;
};
