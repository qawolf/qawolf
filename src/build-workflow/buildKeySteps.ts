import Debug from 'debug';
import { isKeyEvent, isPasteEvent, isSelectAllEvent } from './event';
import { removeShortcutKeyEvents } from './removeShortcutKeyEvents';
import {
  Action,
  Doc,
  ElementEvent,
  KeyEvent,
  PasteEvent,
  Step,
} from '../types';

type BuildStepOptions = {
  action?: Action;
  event: ElementEvent;
  value: string;
};

const debug = Debug('qawolf:buildKeySteps');

const MODIFIER_KEYS = ['Alt', 'Control', 'Meta', 'Shift'];

const isSpecialKey = (key: string): boolean => key.length > 1;

const isModifierKey = (key: string): boolean =>
  MODIFIER_KEYS.some(modifier => key.includes(modifier));

export class KeyStepFactory {
  private _events: ElementEvent[];
  private _pendingKeydownEvents: KeyEvent[] = [];
  private _selectAllTarget: Doc | null = null;
  private _steps: Step[] = [];

  constructor(events: ElementEvent[]) {
    this._events = events;
  }

  private _buildStep({ event, ...options }: BuildStepOptions): void {
    let action: Action = options.action || 'type';

    if (
      !options.action &&
      this._selectAllTarget &&
      JSON.stringify(this._selectAllTarget) === JSON.stringify(event.target)
    ) {
      this._selectAllTarget = null;
      action = 'fill';
    }

    const step: Step = {
      action,
      event,
      index: this._steps.length,
      value: options.value,
    };

    debug('build step %j', step);
    this._steps.push(step);
  }

  private _buildPendingKeydowns(): void {
    const events = this._pendingKeydownEvents;
    if (!events.length) return;

    const value = events.map(e => e.value).join('');
    debug(`build pending keydowns ${events.length}: ${value}`);
    this._buildStep({ event: events[0], value });
    this._pendingKeydownEvents = [];
  }

  private _didPageChange(event: ElementEvent): boolean {
    if (!this._pendingKeydownEvents.length) return false;
    return event.page !== this._pendingKeydownEvents[0].page;
  }

  private _handleKeyDown(event: KeyEvent): void {
    if (isModifierKey(event.value || '')) {
      // ignore modifier keys for now since the modified key is captured by event.key
      // XXX add keyboard.up/down support in https://github.com/qawolf/qawolf/issues/492
      debug(`skip modifier ${event.value}`);
      return;
    }

    if (isSpecialKey(event.value)) {
      this._buildPendingKeydowns();
      debug(`build press ${event.value}`);
      this._buildStep({ action: 'press', event, value: event.value });
    } else {
      this._pendingKeydownEvents.push(event);
    }
  }

  public buildSteps(): Step[] {
    const filteredEvents = removeShortcutKeyEvents(
      'selectall',
      removeShortcutKeyEvents('paste', this._events),
    );

    filteredEvents.forEach(event => {
      if (!event.isTrusted) return;

      if (this._didPageChange(event)) this._buildPendingKeydowns();

      if (event.name === 'keydown') {
        this._handleKeyDown(event as KeyEvent);
        return;
      }

      // when the event is no longer a key event, build pending keydowns
      if (!isKeyEvent(event)) this._buildPendingKeydowns();

      if (isPasteEvent(event)) {
        this._buildStep({ event, value: (event as PasteEvent).value });
      } else if (isSelectAllEvent(event)) {
        this._selectAllTarget = event.target;
      }
    });

    // build the pending keydowns after the last event
    this._buildPendingKeydowns();

    return this._steps;
  }
}

export const buildKeySteps = (events: ElementEvent[]): Step[] => {
  const factory = new KeyStepFactory(events);
  return factory.buildSteps();
};
