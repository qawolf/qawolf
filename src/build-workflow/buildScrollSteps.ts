import { ElementEvent, ScrollEvent, Step } from '../types';

export const buildScrollSteps = (events: ElementEvent[]): Step[] => {
  const steps: Step[] = [];

  for (let i = 0; i < events.length; i++) {
    const event = events[i] as ScrollEvent;

    // ignore other actions
    if (event.name !== 'scroll') continue;

    // ignore system initiated scrolls
    if (!event.isTrusted) continue;

    // skip to the last scroll on this target
    const nextEvent = i + 1 < events.length ? events[i + 1] : null;
    if (nextEvent && nextEvent.name === 'scroll') continue;

    steps.push({
      action: 'scroll',
      event,
      index: steps.length,
      value: event.value,
    });
  }

  return steps;
};
