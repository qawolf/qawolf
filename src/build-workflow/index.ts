export * from './buildClickSteps';
export * from './buildKeySteps';
export * from './buildScrollSteps';
export * from './buildSelectSteps';
export * from './buildSteps';
