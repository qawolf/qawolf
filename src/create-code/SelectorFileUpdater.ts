import { outputJson, remove, pathExists, readJson } from 'fs-extra';
import { ReplContext } from 'playwright-utils';
import { buildSelectors } from '../build-code/buildSelectors';
import { Selectors, Step } from '../types';

type ConstructorOptions = {
  initialSelectors: Selectors;
  path: string;
};

type UpdateOptions = {
  steps: Step[];
};

const loadSelectors = async (path: string): Promise<Selectors> => {
  const codeExists = await pathExists(path);
  if (!codeExists) return {};

  const selectors = await readJson(path);
  return selectors;
};

export class SelectorFileUpdater {
  public static async create(path: string): Promise<SelectorFileUpdater> {
    const initialSelectors = await loadSelectors(path);
    return new SelectorFileUpdater({ path, initialSelectors });
  }

  private _initialSelectors: Selectors;
  private _steps: Step[] = [];
  private _path: string;

  private _lock: boolean;

  protected constructor(options: ConstructorOptions) {
    this._initialSelectors = options.initialSelectors;
    this._path = options.path;
  }

  public async discard(): Promise<void> {
    if (this._initialSelectors.length) {
      await outputJson(this._path, this._initialSelectors, { spaces: ' ' });
    } else {
      await remove(this._path);
    }
  }

  public selectors(): Selectors {
    return {
      ...this._initialSelectors,
      // merge the new selectors with the initial selectors
      // they should not conflict, since we start the new steps at later indices
      ...buildSelectors(this._steps),
    };
  }

  public async update(options: UpdateOptions): Promise<void> {
    this._steps = options.steps;

    if (this._lock) return;
    this._lock = true;

    const updatedSelectors = this.selectors();
    ReplContext.set('selectors', updatedSelectors);
    await outputJson(this._path, updatedSelectors, { spaces: ' ' });

    this._lock = false;
  }
}
