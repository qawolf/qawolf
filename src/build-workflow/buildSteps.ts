import { concat, sortBy } from 'lodash';
import { buildClickSteps } from './buildClickSteps';
import { buildKeySteps } from './buildKeySteps';
import { buildScrollSteps } from './buildScrollSteps';
import { buildSelectSteps } from './buildSelectSteps';
import { ElementEvent, Step } from '../types';

export const buildSteps = (events: ElementEvent[], startIndex = 0): Step[] => {
  const unorderedSteps = concat(
    buildClickSteps(events),
    buildKeySteps(events),
    buildScrollSteps(events),
    buildSelectSteps(events),
  );

  let steps = sortBy(
    unorderedSteps,
    // ordered by the event time
    step => step.event.time,
  );

  // reindex
  steps = steps.map((step, index) => ({
    ...step,
    index: index + startIndex,
  }));

  return steps;
};
