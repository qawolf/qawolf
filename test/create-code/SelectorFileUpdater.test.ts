import { outputJson, pathExists, readJson, remove } from 'fs-extra';
import { buildSteps } from '../../src/build-workflow/buildSteps';
import { SelectorFileUpdater } from '../../src/create-code/SelectorFileUpdater';

// require manually since fs is mocked
// eslint-disable-next-line @typescript-eslint/no-var-requires
const { events } = require('../fixtures/login.json');
const steps = buildSteps(events);

jest.mock('fs-extra');

(pathExists as jest.Mock).mockResolvedValue(true);

const mockedOutputJson = outputJson as jest.Mock;
const mockedReadJson = readJson as jest.Mock;
const mockedRemove = remove as jest.Mock;

describe('SelectorFileUpdater', () => {
  beforeEach(() => {
    mockedOutputJson.mockReset();
  });

  describe('discard', () => {
    it('reverts to initial selectors', async () => {
      const initial = ['css=selector0', 'css=selector1'];
      mockedReadJson.mockResolvedValue(initial);
      const updater = await SelectorFileUpdater.create('revertpath');
      await updater.discard();
      const reverted = mockedOutputJson.mock.calls[0];
      expect(reverted[0]).toEqual('revertpath');
      expect(reverted[1]).toEqual(initial);
      expect(mockedRemove.mock.calls.length).toEqual(0);
    });

    it('removes file if there were no initial selectors', async () => {
      const initial = [];
      mockedReadJson.mockResolvedValue(initial);
      const updater = await SelectorFileUpdater.create('removepath');
      await updater.discard();
      expect(mockedRemove.mock.calls[0][0]).toEqual('removepath');
    });
  });

  describe('update', () => {
    it('appends new selectors', async () => {
      const initial = ['css=selector0'];
      mockedReadJson.mockResolvedValue(initial);

      const updater = await SelectorFileUpdater.create('revertpath');

      // emulate only the first step being ready
      await updater.update({ steps: steps.slice(0, 1) });
      const revisionOne = mockedOutputJson.mock.calls[0][1];
      expect(revisionOne).toMatchSnapshot();
      mockedReadJson.mockResolvedValueOnce(revisionOne);

      await updater.update({ steps: steps.slice(0, 2) });
      const revisionTwo = mockedOutputJson.mock.calls[1][1];
      expect(revisionTwo).toMatchSnapshot();
    });
  });
});
