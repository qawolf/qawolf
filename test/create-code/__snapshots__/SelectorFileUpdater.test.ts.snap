// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectorFileUpdater update appends new selectors 1`] = `
Object {
  "0": "css=selector0",
  "0_a": "html=<ul qaw_innertext=\\"Buttons Checkbox inputs Content editables Date pickers Images Infinite scroll Large Log in Radio inputs Selects Text inputs Time pickers\\"><li qaw_innertext=\\"Log in\\"><a href=\\"/login\\" qaw_innertext=\\"Log in\\"></a></li></ul>",
}
`;

exports[`SelectorFileUpdater update appends new selectors 2`] = `
Object {
  "0": "css=selector0",
  "0_a": "html=<ul qaw_innertext=\\"Buttons Checkbox inputs Content editables Date pickers Images Infinite scroll Large Log in Radio inputs Selects Text inputs Time pickers\\"><li qaw_innertext=\\"Log in\\"><a href=\\"/login\\" qaw_innertext=\\"Log in\\"></a></li></ul>",
  "1_username_input": "html=<div id=\\"root\\" qaw_innertext=\\"To log in, use tomsmith as the username, and SuperSecretPassword! as the password. Username Password Log in\\"><form qaw_innertext=\\"Username Password Log in\\"><input autocomplete=\\"off\\" id=\\"username\\" type=\\"text\\" value=\\"\\" qaw_labels=\\"Username\\" /></form></div>",
}
`;
